File: ts/instructions/setEntanglerState.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface SetEntanglerStateArgs {
  admin: PublicKey
  earner: PublicKey
  price: BN
}

export interface SetEntanglerStateAccounts {
  signer: PublicKey
  feeMint: PublicKey
  /** The entangled collection mint */
  state: PublicKey
  /** Common Solana programs */
  tokenProgram: PublicKey
  systemProgram: PublicKey
  rent: PublicKey
}

export const layout = borsh.struct([
  borsh.publicKey("admin"),
  borsh.publicKey("earner"),
  borsh.u64("price"),
])

/** Sets the state of the entangler */
export function setEntanglerState(
  args: SetEntanglerStateArgs,
  accounts: SetEntanglerStateAccounts
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.feeMint, isSigner: false, isWritable: false },
    { pubkey: accounts.state, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([3, 157, 167, 221, 222, 29, 49, 14])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      admin: args.admin,
      earner: args.earner,
      price: args.price,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/entangler/src/instructions/set_entangler_state.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{Mint, Token};

use crate::seeds::STATE_SEED;
use crate::state::EntanglerState;

pub fn set_entangler_state(
    ctx: Context<SetEntanglerState>,
    admin: Pubkey,
    earner: Pubkey,
    price: u64,
) -> Result<()> {
    msg!("Creating the collection entry");

    let entangler_state = &mut ctx.accounts.state;
    entangler_state.admin = admin;
    entangler_state.earner = earner;
    entangler_state.fee_mint = ctx.accounts.fee_mint.key();
    entangler_state.price = price;

    Ok(())
}

#[derive(Accounts)]
#[instruction(admin: Pubkey)]
pub struct SetEntanglerState<'info> {
    #[account(
        mut,
        constraint = signer.key() == state.admin || state.admin == Pubkey::default(),
    )]
    pub signer: Signer<'info>,

    pub fee_mint: Account<'info, Mint>,

    /// The entangled collection mint
    #[account(
        init_if_needed,
        space = EntanglerState::LEN,
        payer = signer,
        seeds = [
            STATE_SEED.as_bytes(),
        ],
        bump,
    )]
    pub state: Box<Account<'info, EntanglerState>>,

    /// Common Solana programs
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
